utils/report.c

```diff
@@ -81,6 +81,126 @@ static struct uftrace_report_node *find_or_create_node(struct rb_root *root, con
 	return node;
 }
 
+#define DEFAULT_CALL_STAT_SIZE 4
+
+struct call_stat {
+	uint64_t addr;
+	struct rb_node link;
+	uint64_t *callers;
+	uint64_t *callees;
+	int caller_alloc;
+	int callee_alloc;
+	int caller_count;
+	int callee_count;
+};
+
+static struct rb_root call_stat_root = RB_ROOT;
+
+static struct call_stat *find_or_create_call_stat(struct uftrace_fstack *fstack)
+{
+	struct call_stat *call_stat;
+	struct rb_node *parent = NULL;
+	struct rb_node **p = &call_stat_root.rb_node;
+
+	while (*p) {
+		int cmp;
+
+		parent = *p;
+		call_stat = rb_entry(parent, typeof(*call_stat), link);
+
+		cmp = call_stat->addr - fstack->addr;
+		if (cmp == 0)
+			return call_stat;
+
+		if (cmp > 0)
+			p = &parent->rb_left;
+		else
+			p = &parent->rb_right;
+	}
+
+	call_stat = xmalloc(sizeof(*call_stat));
+	call_stat->addr = fstack->addr;
+	call_stat->callers = xmalloc(sizeof(*call_stat->callers) * DEFAULT_CALL_STAT_SIZE);
+	call_stat->callees = xmalloc(sizeof(*call_stat->callees) * DEFAULT_CALL_STAT_SIZE);
+	call_stat->caller_alloc = DEFAULT_CALL_STAT_SIZE;
+	call_stat->callee_alloc = DEFAULT_CALL_STAT_SIZE;
+	call_stat->caller_count = 0;
+	call_stat->callee_count = 0;
+
+	rb_link_node(&call_stat->link, parent, p);
+	rb_insert_color(&call_stat->link, &call_stat_root);
+
+	return call_stat;
+}
```

Comment:
General advice in this code.  I prefer short names for local variables when it's obvious.  In this case we know it's dealing with call_stat so it could be just `cs` or `stat`.  Also caller and callee allocation logic is the same then we can factor out it as a struct like below:

```
struct call_stat_data {
    uint64 *addrs;
    int alloc;
    int count;
};

struct call_stat {
    uint64_t addr;
    struct rb_node link;
    struct call_stat_data caller;
    struct call_stat_data callee;
};
```